README.md
# Simply Hexagonal Logger
![Tests](https://github.com/simplyhexagonal/logger/workflows/tests/badge.svg)
[![Try logger on RunKit](https://badge.runkitcdn.com/@simplyhexagonal/logger.svg)](https://npm.runkit.com/@simplyhexagonal/logger)

Extensible asynchronous debug logger with singleton capabilities, developed to easily broadcast to
multiple communication channels/transports.

```ts
import Logger from '@simplyhexagonal/logger';

const logger = new Logger({});

logger.debug('Trying to teach', 2, 'tooters', {to: 'toot'});
```

## Open source notice

This project is open to updates by its users, [I](https://github.com/jeanlescure) ensure that PRs are relevant to the community.
In other words, if you find a bug or want a new feature, please help us by becoming one of the
[contributors](#contributors-) ✌️ ! See the [contributing section](#contributing)

## Like this module? ❤

Please consider:

- [Buying me a coffee](https://www.buymeacoffee.com/jeanlescure) ☕
- Supporting Simply Hexagonal on [Open Collective](https://opencollective.com/simplyhexagonal) 🏆
- Starring this repo on [Github](https://github.com/simplyhexagonal/logger) 🌟

## Features

- Define different communication channels per log level (i.e. send `debug` messages to console and `error` messages to Slack)
- Define multiple communication channels per log level
- Specify channel names and then simply use the `.channel()` function to send messages to any specific channel
- Extend your logging capabilities with officially supported transports for: Slack, Discord, Email, SMS, Socket
- Easily make your own transports by implementing and extending the [base LoggerTransport class type](https://github.com/simplyhexagonal/logger/blob/main/src/transports/base.ts) (and submit them via GitHub issue for adoption as an officially supported transport!)
- Use the same logger instance throughout your app (singleton logger)
- Use multiple logger instances throughout your app with de-duplicated communication channels/transports (singleton transports)
- ANSI colors and easy to read formatting for CLI terminals
- CSS colors and easy to read formatting for browser dev consoles

## Usage

Install:

```sh
npm install @simplyhexagonal/logger

yarn add @simplyhexagonal/logger

pnpm i @simplyhexagonal/logger
```

There are three basic configuration elements you should established based on your app's needs and
the environment you will deploy to:

- log level
- communication channels
- error management strategy

### Log level

There are 6 log levels:

- debug
- info
- warn
- error
- fatal
- all

And a bypass level that outputs the message with no date, level, nor colors:

- raw

In your code you log messages to specific log levels:

```ts
logger.debug('hello');

try {
  throw new Error('you shall not pass');
} catch {
  logger.error('time to turn around');
}

// 2021-10-02T23:47:27.187Z DEBUG 🐞️:
//
//    hello
//

// 2021-10-02T23:47:27.191Z ERROR 🚨️:
//
//    time to turn around
//
```

Each log level is given a number value:

```ts
{
  debug: 0,
  info: 10,
  warn: 20,
  error: 30,
  fatal: 40,
  all: 100,
  raw: 110,
}
```

When `Logger` is instantiated, it will only setup communication channels for the configured log
level and above (i.e. if you selected `warn` then the only `logLevel >= 20` would be initialized)

You can set the log level when instancing the logger:

```ts
import Logger, { LogLevels } from '@simplyhexagonal/logger';

new Logger({
  logLevel: LogLevels.DEBUG,
  //...
});
```

It is **highly recommended** to set the log level based on a condition that determines the environment
your app is running on:

```ts
const logLevel = (process.env.NODE_ENV === 'production') ? LogLevels.ERROR : LogLevels.DEBUG;

new Logger({
  logLevel,
  //...
});
```

Setting the log level using environment variables is only recommended as a way to override the
log level configured during instantiation:

```sh
# .env
LOG_LEVEL=debug
```

_(i.e. this is useful if you deem it necessary to turn on debug logging in production environments)_

### Communication channels

Let's say that you have a Discord server with a channel you want to receive only debug messages from
your app, and another channel dedicated to receiving only errors.

The debug channel has the webhook path: `/D3BU9/W3BH00K`

The error channel has the webhook path: `/3RR0R/W3BH00K`

Using `@simplyhexagonal/logger`, you can add the [official Discord transport](https://www.npmjs.com/package/@simplyhexagonal/logger-transport-discord) as a dependency
and import it:

```ts
import DiscordTransport from '@simplyhexagonal/logger-transport-discord';
```

Then, you can configure the transports for each of Logger's log levels:

```ts
const optionsByLevel = {
  debug: [
    {
      transport: LoggerTransportName.DISCORD,
      options: {
        // debug channel webhook url
        destination: 'https://discord.com/api/webhooks/D3BU9/W3BH00K',
      },
    },
  ],
  info: [],
  warn: [],
  error: [
    {
      transport: LoggerTransportName.DISCORD,
      options: {
        // error channel webhook url
        destination: 'https://discord.com/api/webhooks/3RR0R/W3BH00K',
      },
    },
  ],
  fatal: [],
  all: [],
};
```

Then you would let Logger know which transport to use for `LoggerTransportName.DISCORD`:

```ts
const transports = {
  [LoggerTransportName.DISCORD]: DiscordTransport,
};
```

The final result would look something like this:

```ts
import {
  LogLevels,
  LoggerTransportName,
} from '@simplyhexagonal/logger';
import DiscordTransport from '@simplyhexagonal/logger-transport-discord';

const optionsByLevel = {
  debug: [
    {
      transport: LoggerTransportName.DISCORD,
      options: {
        // debug channel webhook url
        destination: 'https://discord.com/api/webhooks/D3BU9/W3BH00K',
      },
    },
  ],
  info: [],
  warn: [],
  error: [
    {
      transport: LoggerTransportName.DISCORD,
      options: {
        // error channel webhook url
        destination: 'https://discord.com/api/webhooks/3RR0R/W3BH00K',
      },
    },
  ],
  fatal: [],
  all: [],
  raw: [],
};

const transports = {
  [LoggerTransportName.DISCORD]: DiscordTransport,
};

const options = {
  logLevel: LogLevels.DEBUG,
  optionsByLevel,
  transports,
};

const logger = new Logger(options);
```

### Error management strategy

In the previous example there's always a possibility for the Discord webhook to return an error.

When this happens Logger will default to throwing an error which can be handled using `.catch()`:

```ts
logger.debug('hello discord').catch((e) => {
  logger.channel(LoggerTransportName.CONSOLE).error(e);
});
```

We understand that this adds unnecessary complexity, as such, you are encouraged to turn on error
catching when instantiating `Logger`. When you do this, Logger will automagically catch transport
errors and log them to console (with `error` log level):

```ts
const options = {
  optionsByLevel: optionsWithBadTransport,
  catchTransportErrors: true,
};

const logger = new Logger(options);

logger.debug('this will fail due to a bad transport');

//  2021-10-03T04:31:02.191Z ERROR 🚨️:
//  
//      {
//        "transportResult": {
//          "destination": "...",
//          "channelName": "...",
//          "error": {
//            "name": "Error",
//            "message": "LOGGER ERROR: ...",
//            "stack": "Error: LOGGER ERROR: ...",
//            // ...
//          },
//        },
//        //...
//      }
//
```

Furthermore, you could implement your own fallback transport:

```ts
import { LoggerTransport } from '@simplyhexagonal/logger/transports/base';

class MyTransport extends LoggerTransport {
  constructor(options: LoggerTransportOptions['options']) {
    const r = Math.random().toString(36).substring(7);
    super({...options, r});
  }

  async error([timestamp, ...message]: unknown[]) {
    console.log(timestamp, 'MY LOG:', ...message);

    return {
      destination: this.destination,
      channelName: this.channelName,
    };
  }
}

const options = {
  optionsByLevel: optionsWithBadTransport,
  catchTransportErrors: true,
  fallbackTransport: MyTransport,
};

const logger = new Logger(options);

logger.debug('this will fail due to a bad transport');

// 2021-10-03T04:31:02.201Z MY LOG: UndefinedTransportError: ...
```

And just as with `LOG_LEVEL`, we have implemented an environment variable for overriding purposes:

```sh
# .env
LOGGER_CATCH_TRANSPORT_ERRORS=true
```

**IMPORTANT NOTE:** we recommend always setting `catchTransportErrors` to `true` in production!

### More options

```ts
import {
  LogLevels,
  LoggerTransportName,
} from '@simplyhexagonal/logger';
import DiscordTransport from '@simplyhexagonal/logger-transport-discord';

const options = {
  logLevel: LogLevels.DEBUG, // default
  optionsByLevel: {
    debug: [
      // ***
      // This console config is the default if a log level options array is left empty
      // (like `info` in this example)
      {
        transport: LoggerTransportName.CONSOLE,
        options: {
          destination: LoggerTransportName.CONSOLE,
          channelName: LoggerTransportName.CONSOLE,
        },
      },
      // if you do this you would have only one instance of this transport since all
      // transports are singleton (as in pre-filtered and de-duplicated)
      // ***
      {
        transport: LoggerTransportName.DISCORD,
        options: {
          destination: 'https://discord.com/api/webhooks/D3BU9/W3BH00K',
          channelName: 'discord-debug',
        },
      },
    ],
    info: [], // in this case `loggger.info()` will default to logging to the console
    warn: [],
    error: [
      {
        transport: LoggerTransportName.DISCORD,
        options: {
          destination: 'https://discord.com/api/webhooks/3RR0R/W3BH00K',
        },
      },
    ],
    fatal: [],
    all: [],
    raw: [],
  },
  transports: {
    [`${LoggerTransportName.DISCORD}`]: DiscordTransport,
  },
  singleton: true, // default
  catchTransportErrors: false, // default
  fallbackTransport: MyTransport,
};

const logger = new Logger(options);
```

## The `all` and `raw` log levels

An important thing to note is that transports defined for the `all` and `raw` log levels will
**always** be instantiated.

In the same way, calls to `logger.all()` or `logger.raw()` will **always** log.

For this reason we suggest only ever using `logger.all()` when an app starts and when an app is
manually stopped, and only use `logger.raw()` when outputting messages you absolutely **need** to
be unformatted.

## Channels

It can be extremely useful to setup multiple channels for specific purposes on a log level:

```ts
const logger = new Logger({
  logLevel: LogLevels.DEBUG,
  optionsByLevel: {
    warn: [],
    info: [],
    debug: [],
    error: [
      {
        transport: LoggerTransportName.SLACK,
        options: {
          destination: 'https://hooks.slack.com/services/T123/B456/M0N90',
          channelName: 'mongo',
        },
      },
      {
        transport: LoggerTransportName.SLACK,
        options: {
          destination: 'https://hooks.slack.com/services/T123/B456/M55QL',
          channelName: 'mssql',
        },
      },
    ],
    fatal: [],
    all: [],
  },
});
```

and then send messages to a specific channel depending on the event that's triggering the log:

```ts
const server = async () => {
  await mongoose.connect('mongodb://mymongo.cluster:27017/myapp').catch((e) => {
    logger.channel('mongo').error(e);
  });

  await sql.connect('Server=mymssql.cluster,1433;Database=myapp;').catch((e) => {
    logger.channel('mssql').error(e);
  });
}

server();
```

## Transports

We have the following officially supported transports:

- [WebSocket]() (Coming soon)
- [Discord](https://www.npmjs.com/package/@simplyhexagonal/logger-transport-discord)
- [Slack]() (Coming soon)
- [Email]() (Coming soon)
- [SMS]() (Coming soon)

## Time functions to measure performance

We have implemented functions similar to `console.time()` and `console.timeEnd()` to measure performance:

```ts
const logger = new Logger({});

logger.time('my-timed-operation');

// do things

const result: number = await logger.timeEnd('my-timed-operation');

// transport output:
//
//   my-timed-operation: 123.456 ms
//

console.log(result);

// 123.456
```

Three important differences from the console functions to note:

- `logger.timeEnd` is async as it logs to the transports through the `.raw()` method
- `logger.timeEnd` will return a number representing the elapsed time in milliseconds, which is useful for being able to store the value for usage in measuring performance through code
- we used `performance.timeOrigin + performance.now()` instead of `Date.now()` to get the current time in high resolution milliseconds, due to this fact the time measurements are more accurate than some internal time-based functions (e.g. from experience we've seen code like `setTimeout(() => {}, 100)` might be measured to take `99.83 ms` which can break a test if you expect something like `await logger.timeEnd('...') > 100`)

## Contributing

Yes, thank you! This plugin is community-driven, most of its features are from different authors.
Please update the docs and tests and add your name to the `package.json` file.

## Contributors ✨

Thanks goes to these wonderful people ([emoji key](https://allcontributors.org/docs/en/emoji-key)):

<!-- ALL-CONTRIBUTORS-LIST:START - Do not remove or modify this section -->
<!-- prettier-ignore-start -->
<!-- markdownlint-disable -->
<table>
  <tr>
    <td align="center"><a href="https://jeanlescure.cr"><img src="https://avatars2.githubusercontent.com/u/3330339?v=4" width="100px;" alt=""/><br /><sub><b>Jean Lescure</b></sub></a><br /><a href="#maintenance-jeanlescure" title="Maintenance">🚧</a> <a href="https://github.com/simplyhexagonal/logger/commits?author=jeanlescure" title="Code">💻</a> <a href="#userTesting-jeanlescure" title="User Testing">📓</a> <a href="https://github.com/simplyhexagonal/logger/commits?author=jeanlescure" title="Tests">⚠️</a> <a href="#example-jeanlescure" title="Examples">💡</a> <a href="https://github.com/simplyhexagonal/logger/commits?author=jeanlescure" title="Documentation">📖</a></td>
    <td align="center"><a href="https://github.com/2huBrulee"><img src="https://avatars.githubusercontent.com/u/29010617?v=4" width="100px;" alt=""/><br /><sub><b>Alejandro Merino</b></sub></a><br /><a href="#maintenance-2huBrulee" title="Maintenance">🚧</a> <a href="https://github.com/simplyhexagonal/logger/commits?author=2huBrulee" title="Code">💻</a> <a href="#userTesting-2huBrulee" title="User Testing">📓</a> <a href="https://github.com/simplyhexagonal/logger/commits?author=2huBrulee" title="Tests">⚠️</a></td>
  </tr>
</table>

<!-- markdownlint-enable -->
<!-- prettier-ignore-end -->
<!-- ALL-CONTRIBUTORS-LIST:END -->
## License

Copyright (c) 2021-Present [Logger Contributors](https://github.com/simplyhexagonal/logger/#contributors-).<br/>
Licensed under the [Apache License 2.0](https://www.apache.org/licenses/LICENSE-2.0).



CHANGELOG.md
### Changelog

All notable changes to this project will be documented in this file. Dates are displayed in UTC.

Generated by [`auto-changelog`](https://github.com/CookPete/auto-changelog).

#### [2.1.1](https://github.com/simplyhexagonal/logger/compare/2.1.0...2.1.1)

- chore: build dist for release [`2a4253b`](https://github.com/simplyhexagonal/logger/commit/2a4253b531afbbd3701f7c829b8278c166522c40)
- chore: add defaultception [`947ced1`](https://github.com/simplyhexagonal/logger/commit/947ced1f8d97f280ef7a9cd9cbe037906edd7646)

#### [2.1.0](https://github.com/simplyhexagonal/logger/compare/2.0.0...2.1.0)

> 15 February 2022

- chore: use matrix.os on test github workflow [`6519f4c`](https://github.com/simplyhexagonal/logger/commit/6519f4cde2790b633394494c624d79600a442b36)
- docs: update README to include new time functions [`d82f6fe`](https://github.com/simplyhexagonal/logger/commit/d82f6fe1a0506acb0f43edcd7027aed7fa93ab51)
- chore: build dist for release [`6741119`](https://github.com/simplyhexagonal/logger/commit/67411194ad529dd7d818235daa3f198c927414f0)

### [2.0.0](https://github.com/simplyhexagonal/logger/compare/1.3.0...2.0.0)

> 10 January 2022

- chore: build dist for release [`52866c4`](https://github.com/simplyhexagonal/logger/commit/52866c4e27454274f64c82c05e79169fb1b3f4ae)
- fix: make perf_hooks external in build script [`e35de54`](https://github.com/simplyhexagonal/logger/commit/e35de5435c0defac889a31c33b26047693a7c600)
- chore: uncomment tests [`fac03db`](https://github.com/simplyhexagonal/logger/commit/fac03dba2f1cc8ce6d7d1e85481e8a43e30330c4)

#### [1.3.0](https://github.com/simplyhexagonal/logger/compare/1.2.2...1.3.0)

> 14 December 2021

- feat: add ability to output 'raw' messages [`0405137`](https://github.com/simplyhexagonal/logger/commit/040513713aeb5b3a86fcdedae37e1c01b6added1)
- chore: build dist for release [`fd89c77`](https://github.com/simplyhexagonal/logger/commit/fd89c77d356fbfcb2113a92bb4a8fdf35f22ab7f)
- Release 1.3.0 [`c72e61e`](https://github.com/simplyhexagonal/logger/commit/c72e61e543e7b612d481e1f73b1d4de31cf5193d)

#### [1.2.2](https://github.com/simplyhexagonal/logger/compare/1.2.1...1.2.2)

> 15 October 2021

- chore: build dist for release [`96bd6a4`](https://github.com/simplyhexagonal/logger/commit/96bd6a42e68d67c4ba1f796e0df80b95ec3b934a)
- chore: refactor timestamp to prefixes [`2f7994c`](https://github.com/simplyhexagonal/logger/commit/2f7994cb27033d9ce9562b5a16f20e78b7577bb6)
- chore: bundle only deps that do not have browser versions [`178003c`](https://github.com/simplyhexagonal/logger/commit/178003c426fa57c8143d94a12f5a85065cb3ccfd)

#### [1.2.1](https://github.com/simplyhexagonal/logger/compare/1.2.0...1.2.1)

> 14 October 2021

- Release 1.2.1 [`3d4849e`](https://github.com/simplyhexagonal/logger/commit/3d4849e00797603e3f10730c3c051df44c3d9413)
- chore: build dist for release [`f1a158c`](https://github.com/simplyhexagonal/logger/commit/f1a158c08c48fa0f8ab001661f790d3588a43d61)
- fix: missing @types/xxhashjs for transport development [`ed443d8`](https://github.com/simplyhexagonal/logger/commit/ed443d8883db3f46757839167beab45a2b815092)

#### [1.2.0](https://github.com/simplyhexagonal/logger/compare/1.1.2...1.2.0)

> 14 October 2021

- chore: build dist for release [`f003073`](https://github.com/simplyhexagonal/logger/commit/f00307396fa1e4564625dc6f2d617350cd54c5e3)
- chore: refactor appIdentifier/date formatting [`54370f5`](https://github.com/simplyhexagonal/logger/commit/54370f5bb7dd0cd86747d56361ce96a9f61382f5)
- Release 1.2.0 [`6d366cc`](https://github.com/simplyhexagonal/logger/commit/6d366cc7b143363dc7ecace63a03f13d396f4003)

#### [1.1.2](https://github.com/simplyhexagonal/logger/compare/1.1.1...1.1.2)

> 14 October 2021

- chore: build dist for release [`5545d14`](https://github.com/simplyhexagonal/logger/commit/5545d1460f6d69f7140fec2133c4bb6b6599ca54)
- Release 1.1.2 [`349f8ab`](https://github.com/simplyhexagonal/logger/commit/349f8abc66ad82a4bc21902e4f2671744cc7fb5e)
- fix: Class extends value undefined error [`b6056a2`](https://github.com/simplyhexagonal/logger/commit/b6056a24ad1e4995ec4739542a6f692b614854d9)

#### [1.1.1](https://github.com/simplyhexagonal/logger/compare/1.1.0...1.1.1)

> 14 October 2021

- Release 1.1.1 [`bb87207`](https://github.com/simplyhexagonal/logger/commit/bb872072b76ea72f57fcd9148b52386e79a6a794)
- chore: build dist for release [`00a520f`](https://github.com/simplyhexagonal/logger/commit/00a520fdd85e0793588a747fb99071d1910422e2)
- fix: CONSOLE is missing in type error [`c3136d8`](https://github.com/simplyhexagonal/logger/commit/c3136d8a0bbd0044d297523cf22c591b32d80128)

#### [1.1.0](https://github.com/simplyhexagonal/logger/compare/1.0.2...1.1.0)

> 14 October 2021

- feat: add ability to configure app identifiers [`2ae171a`](https://github.com/simplyhexagonal/logger/commit/2ae171a1514b3a33d513b91714939f0b50e9aef0)
- chore: build dist for release [`3a92a78`](https://github.com/simplyhexagonal/logger/commit/3a92a7812645554e3c2142858816b0f0e78a4791)
- Release 1.1.0 [`9afc155`](https://github.com/simplyhexagonal/logger/commit/9afc155a58392698b31738f0247da1d20ff2b632)

#### [1.0.2](https://github.com/simplyhexagonal/logger/compare/1.0.1...1.0.2)

> 14 October 2021

- chore: build dist for release [`53c2378`](https://github.com/simplyhexagonal/logger/commit/53c23780a0f5e5763c5bef2657212916aa38d22d)
- chore: streamline interface, base transport, and enum exports [`8b5d419`](https://github.com/simplyhexagonal/logger/commit/8b5d419728b5549cf6caad67aeb24bbdb33f7c02)
- Release 1.0.2 [`34d4f5a`](https://github.com/simplyhexagonal/logger/commit/34d4f5ab42674003c4151bbe84d93ae39c6b25d4)

#### [1.0.1](https://github.com/simplyhexagonal/logger/compare/1.0.0...1.0.1)

> 6 October 2021

- Release 1.0.1 [`301eb97`](https://github.com/simplyhexagonal/logger/commit/301eb977486ccbdd5a2d803320dfb385cceddd33)

src/index.ts
// @ts-ignore
import { version } from '../package.json';

import {
  LoggerTransportNameEnum,
  LogLevelsEnum,
  TransportInstances,
  LoggerTransportOptions,
  LoggerTransportOptionsByLevel,
  LoggerTransportResult,
  LoggerBroadcastFns,
  AppIdentifiers,
} from './interfaces';

import { LoggerTransport } from './transports/base';
import ConsoleTransport from './transports/console';
import UndefinedTransport from './transports/undefined';

export * from './interfaces';

export { LoggerTransport } from './transports/base';

export const LoggerTransportName = {...LoggerTransportNameEnum};
export const LogLevels = {...LogLevelsEnum};

export interface LoggerTransportClasses {
  [LoggerTransportName.CONSOLE]: typeof LoggerTransport;
  [LoggerTransportName.SLACK]?: typeof LoggerTransport;
  [LoggerTransportName.DISCORD]?: typeof LoggerTransport;
  [LoggerTransportName.EMAIL]?: typeof LoggerTransport;
  [LoggerTransportName.SMS]?: typeof LoggerTransport;
  [LoggerTransportName.SOCKET]?: typeof LoggerTransport;
  [k: string]: typeof LoggerTransport | undefined;
}

let performanceShim: Performance;

if (typeof window === 'undefined' || !window.performance) {
  performanceShim = require('perf_hooks').performance;
} else {
  performanceShim = window.performance;
}

const LOG_LEVELS = {
  debug: 0,
  info: 10,
  warn: 20,
  error: 30,
  fatal: 40,
  all: 100,
  raw: 110,
};

export interface LoggerOptions {
  optionsByLevel?: LoggerTransportOptionsByLevel;
  transports?: Partial<LoggerTransportClasses>;
  singleton?: boolean;
  logLevel?: keyof typeof LOG_LEVELS;
  catchTransportErrors?: boolean;
  fallbackTransport?: typeof LoggerTransport;
  appIdentifiers?: AppIdentifiers;
}

const defaultLoggerTransportOptions: LoggerTransportOptions = {
  transport: LoggerTransportName.CONSOLE,
  options: {
    destination: LoggerTransportName.CONSOLE,
    channelName: LoggerTransportName.CONSOLE,
  },
};

const defaultOptionsByLevel: LoggerTransportOptionsByLevel = {
  debug: [defaultLoggerTransportOptions],
  info: [defaultLoggerTransportOptions],
  warn: [defaultLoggerTransportOptions],
  error: [defaultLoggerTransportOptions],
  fatal: [defaultLoggerTransportOptions],
  all: [defaultLoggerTransportOptions],
  raw: [defaultLoggerTransportOptions],
};

const initialTransportInstances: TransportInstances = {
  debug: [],
  info: [],
  warn: [],
  error: [],
  fatal: [],
  all: [],
  raw: [],
};

const defaultTransports = {
  [LoggerTransportName.CONSOLE]: ConsoleTransport,
  [LoggerTransportName.SLACK]: UndefinedTransport,
  [LoggerTransportName.DISCORD]: UndefinedTransport,
  [LoggerTransportName.EMAIL]: UndefinedTransport,
  [LoggerTransportName.SMS]: UndefinedTransport,
  [LoggerTransportName.SOCKET]: UndefinedTransport,
};

/**
 * The `Logger` class does the heavy-lifting of keeping track of transports,
 * broadcasting to the appropriate transports (or not) based on the log levels,
 * and aggregating the results from said broadcasts.
 */
export default class Logger {
  private static instance: Logger;

  static version: string = version;
  static LoggerTransportName = LoggerTransportName;
  static LogLevels = LogLevels;
  static LoggerTransport = LoggerTransport;
  static hrTime() {
    return performanceShim.timeOrigin + performanceShim.now();
  }

  static default: typeof Logger = Logger;

  _timers: {[k: string]: number} = {};

  optionsByLevel: LoggerTransportOptionsByLevel;
  availableTransports: LoggerTransportClasses;
  transportInstances: TransportInstances = JSON.parse(
    JSON.stringify(initialTransportInstances)
  );
  appIdentifiers: AppIdentifiers = {};

  catchTransportErrors: boolean = false;
  readonly fallbackTransport: undefined | LoggerTransport;

  private appIdString(): string {
    const {
      region,
      clusterType,
      cluster,
      hostname,
      ip,
      app,
    } = this.appIdentifiers;

    const result = [
      region,
      clusterType,
      cluster,
      hostname,
      ip,
      app
    ].filter(
      (i) => i
    ).join(' > ');

    return (result) ? `[${result}]` : '';
  }

  constructor({
    optionsByLevel,
    transports,
    singleton = true,
    logLevel,
    catchTransportErrors,
    fallbackTransport,
    appIdentifiers,
  }: LoggerOptions) {
    this.optionsByLevel = { ...defaultOptionsByLevel, ...optionsByLevel };
    this.availableTransports = {...defaultTransports, ...transports};
    this.appIdentifiers = appIdentifiers || {};

    const {
      LOG_LEVEL,
      LOGGER_CATCH_TRANSPORT_ERRORS,
    } = process.env;

    this.catchTransportErrors = (
      (LOGGER_CATCH_TRANSPORT_ERRORS || '').toLowerCase() === 'true'
    ) || catchTransportErrors || this.catchTransportErrors;

    if (singleton && Logger.instance) return Logger.instance;

    if (singleton && !Logger.instance) Logger.instance = this;

    if (this.catchTransportErrors) {
      if (fallbackTransport) {
        this.fallbackTransport = new fallbackTransport({
          destination: `Logger:Fallback(${fallbackTransport.name})`
        });
      } else {
        this.fallbackTransport = new ConsoleTransport({destination: 'Logger:Fallback'});
      }
    }

    const envLogLevel = LOG_LEVELS[LOG_LEVEL as keyof typeof LOG_LEVELS];

    const destinations: string[] = [];
    const transportInstanceMap: {[k: string]: LoggerTransport} = {};

    /**
     * To save on resources, register singleton transports only for the log levels
     * above or equal the configured log level
     */
    Object.entries(LOG_LEVELS).forEach(([k, v]) => {this.broadcast
      if (
        v >= (
          envLogLevel // This is our environment variable override feature
          || (
            (
              logLevel
            ) ? (
              LOG_LEVELS[logLevel as keyof typeof LOG_LEVELS]
            ) : (
              LOG_LEVELS[LogLevels.DEBUG]
            )
          )
        ) || k === LogLevels.ALL
      ) {
        let lvlOptions = this.optionsByLevel[k as keyof LoggerTransportOptionsByLevel];

        if (lvlOptions.length === 0) {
          lvlOptions = defaultOptionsByLevel[k as keyof LoggerTransportOptionsByLevel];
        }

        lvlOptions.forEach(({transport, options}) => {
          const {destination} = options;
          const TransportClass = this.availableTransports[transport as LoggerTransportNameEnum]
              || UndefinedTransport;

          if (!destinations.includes(destination)) {
            destinations.push(destination);

            transportInstanceMap[destination] = new TransportClass({
              name: transport,
              ...options,
            });
          }

          this.transportInstances[k as keyof TransportInstances].push(
            transportInstanceMap[destination]
          );
        });
      }
    });
  }

  time(operationIdentifier: string) {
    this._timers[operationIdentifier] = Logger.hrTime();
  }

  async timeEnd(operationIdentifier: string) {
    const {
      [`${operationIdentifier}`]: time,
      ...rest
    } = this._timers;

    if (!time) {
      await this.channel(LoggerTransportName.CONSOLE).warn(`Timer '${operationIdentifier}' does not exist`);
      return -1;
    }

    this._timers = rest;

    const result = Logger.hrTime() - time;

    await this.raw(`${operationIdentifier}: ${result} ms`);

    return result;
  }

  cleanupTimers() {
    this._timers = {};
  }

  async debug(...message: unknown[]) {
    return await this.broadcast(message, LogLevels.DEBUG);
  }

  async info(...message: unknown[]) {
    return await this.broadcast(message, LogLevels.INFO);
  }

  async warn(...message: unknown[]) {
    return await this.broadcast(message, LogLevels.WARN);
  }

  async error(...message: unknown[]) {
    return await this.broadcast(message, LogLevels.ERROR);
  }

  async fatal(...message: unknown[]) {
    return await this.broadcast(message, LogLevels.FATAL);
  }

  async all(...message: unknown[]) {
    return await this.broadcast(message, LogLevels.ALL);
  }

  // Convenience alias to make it easier to use the logger
  // as drop-in replacement for console.log
  async log(...message: unknown[]) {
    return await this.broadcast(message, LogLevels.ALL);
  }

  async raw(...message: unknown[]) {
    return await this.broadcast(message, LogLevels.RAW);
  }

  channel(channelName: string): LoggerBroadcastFns {
    return {
      [LogLevels.DEBUG]: async (...message: unknown[]) => {
        return await this.broadcast(message, LogLevels.DEBUG, channelName)
      },
      [LogLevels.INFO]: async (...message: unknown[]) => {
        return await this.broadcast(message, LogLevels.INFO, channelName)
      },
      [LogLevels.WARN]: async (...message: unknown[]) => {
        return await this.broadcast(message, LogLevels.WARN, channelName)
      },
      [LogLevels.ERROR]: async (...message: unknown[]) => {
        return await this.broadcast(message, LogLevels.ERROR, channelName)
      },
      [LogLevels.FATAL]: async (...message: unknown[]) => {
        return await this.broadcast(message, LogLevels.FATAL, channelName)
      },
      [LogLevels.ALL]: async (...message: unknown[]) => {
        return await this.broadcast(message, LogLevels.ALL, channelName)
      },
      [LogLevels.RAW]: async (...message: unknown[]) => {
        return await this.broadcast(message, LogLevels.RAW, channelName)
      },
    };
  }

  async broadcast(
    message: unknown[],
    level: LogLevelsEnum,
    channelName: string = '-',
  ): Promise<LoggerTransportResult[]> {
    const results: Promise<LoggerTransportResult>[] = [];

    await this.transportInstances[level as keyof TransportInstances].reduce(
      async (a, transport) => {
        await a;

        if (channelName === '-' || transport.channelName === channelName) {
          const result = transport[level as LogLevelsEnum](
            [
              [
                this.appIdString(),
                `[${new Date().toISOString()}]`,
              ].join(' '),
              ...message,
            ].filter((m) => m)
          ).catch((e) => {
            if (!this.catchTransportErrors) {
              throw e;
            };

            const fallbackResult = (this.fallbackTransport as LoggerTransport).error(
              [
                [
                  this.appIdString(),
                  `[${new Date().toISOString()}]`,
                ].join(' '),
                e,
              ]
            ).then((r) => ({
              ...(e.transportResult || {}),
              ...r,
            }));

            return fallbackResult;
          });

          results.push(result);

          return result;
        }

        return Promise.resolve();
      },
      Promise.resolve() as Promise<void | LoggerTransportResult>
    );

    return await Promise.all(results);
  }
}
